### PR TITLE
Add animation unit tests

### PR DESCRIPTION
**Pull request by ralphweng2023**
## Description
Adds comprehensive automated unit tests for animation functions in SplashKit. The existing `test_animation.cpp` provides manual/interactive testing, but lacked automated unit tests that can be run as part of CI/CD pipelines.

This PR adds `unit_test_animation.cpp` which provides automated tests for:

**Animation Script functions:**
- `load_animation_script`, `free_animation_script`, `free_all_animation_scripts`
- `has_animation_script`, `animation_script_named`, `animation_script_name`
- `animation_count`, `has_animation_named`, `animation_index`

**Animation functions:**
- `create_animation` (multiple overloads with/without sound)
- `free_animation`, `assign_animation` (multiple overloads)
- `animation_name`, `animation_current_cell`, `animation_ended`
- `animation_entered_frame`, `animation_frame_time`, `animation_current_vector`
- `update_animation`, `restart_animation`

**Edge case handling:**
- Null/invalid pointer handling for all getter functions
- Invalid script file loading
- Animation lifecycle testing (create → update → end → restart)

The tests use the existing `kermit.txt` animation script resource and follow the same patterns as other unit tests in the project.

## Type of change
- [x] New tests

## How Has This Been Tested?
Built the test project using CMake as per build instructions. All tests pass successfully.

**Automated Testing:**
```bash
cd projects/cmake
cmake --preset macOS
cmake --build build/

cd ../../bin
./skunit_tests "[animation],[animation_script]" --success
# Result: All tests passed (86 assertions in 9 test cases)

./skunit_tests --order rand
# Result: All tests passed (1349 assertions in 88 test cases)
```

## Testing Checklist
- [x] Code compiles without errors
- [x] All 9 new animation test cases pass (86 assertions)
- [x] All existing tests still pass (88 total test cases, 1349 assertions)
- [x] Tests run independently in random order

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New tests have been added to prove the functionality works

### CODE REVIEW

File: coresdk/src/test/unit_tests/unit_test_animation.cpp [hunk 1]
```diff
@@ -0,0 +1,374 @@
+/**
+ * Animation Unit Tests
+ *
+ * Tests for animation script and animation functions in SplashKit.
+ * Covers loading, creation, updating, and lifecycle of animations.
```
Comment (rory-cd) [inline coresdk/src/test/unit_tests/unit_test_animation.cpp:5]: Optional: Remove these lines for consistency - other unit tests do not contain a description.

File: coresdk/src/test/unit_tests/unit_test_animation.cpp [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+    REQUIRE(script != nullptr);
+
+    SECTION("can get animation script name")
+    {
+        REQUIRE(animation_script_name(script) == TEST_SCRIPT_NAME);
+    }
+    SECTION("can get animation count")
+    {
+        int count = animation_count(script);
+        REQUIRE(count > 0);
```
Comment (rory-cd) [inline coresdk/src/test/unit_tests/unit_test_animation.cpp:80]: To make this assertion more robust, it could be more specific, e.g. `REQUIRE(count == 14);`

File: coresdk/src/test/unit_tests/unit_test_animation.cpp [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+    {
+        REQUIRE(has_animation_named(script, "WalkFront"));
+        REQUIRE(has_animation_named(script, "walkfront"));  // Case insensitive
+        REQUIRE(has_animation_named(script, "Dance"));
+        REQUIRE_FALSE(has_animation_named(script, "NonExistentAnimation"));
+    }
+    SECTION("can get animation index by name")
+    {
+        int idx = animation_index(script, "WalkFront");
+        REQUIRE(idx >= 0);
```
Comment (rory-cd) [inline coresdk/src/test/unit_tests/unit_test_animation.cpp:92]: To make this assertion more robust, it could be more specific, for example:
```cpp
int idx = animation_index(script, "Dance");
REQUIRE(idx == 8);
```

File: coresdk/src/test/unit_tests/unit_test_animation.cpp [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+
+    SECTION("can get animation name")
+    {
+        std::string name = animation_name(anim);
+        REQUIRE_FALSE(name.empty());
+    }
+    SECTION("can get current cell")
+    {
+        int cell = animation_current_cell(anim);
+        REQUIRE(cell >= 0);
```
Comment (rory-cd) [inline coresdk/src/test/unit_tests/unit_test_animation.cpp:195]: To make this assertion more robust, it could be more specific, e.g. REQUIRE(cell == 0);

File: coresdk/src/test/unit_tests/unit_test_animation.cpp [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+        REQUIRE_FALSE(animation_ended(anim));
+    }
+    SECTION("animation entered frame initially")
+    {
+        REQUIRE(animation_entered_frame(anim));
+    }
+    SECTION("can get frame time")
+    {
+        float frame_time = animation_frame_time(anim);
+        REQUIRE(frame_time >= 0.0f);
```
Comment (rory-cd) [inline coresdk/src/test/unit_tests/unit_test_animation.cpp:208]: To make this section more robust, it could have an additional assertion, e.g. `REQUIRE(frame_time <= 12.0f);`. This would place an upper bound on the frame time.

File: coresdk/src/test/unit_tests/unit_test_animation.cpp [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+    }
+    SECTION("can get current vector")
+    {
+        vector_2d vec = animation_current_vector(anim);
+        // Vector may be (0,0) for static frames
+        REQUIRE((vec.x == vec.x));  // Just check it's a valid number (not NaN)
```
Comment (rory-cd) [inline coresdk/src/test/unit_tests/unit_test_animation.cpp:214]: This line is not wrong, but is difficult to read, and only checks _x_. Perhaps It could be replaced with:
```cpp
REQUIRE(std::isfinite(vec.x));
REQUIRE(std::isfinite(vec.y));
```

File: coresdk/src/test/unit_tests/unit_test_animation.cpp [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+{
+    animation_script script = load_animation_script(TEST_SCRIPT_NAME, TEST_SCRIPT_FILE);
+    REQUIRE(script != nullptr);
+
+    SECTION("animation updates correctly")
+    {
+        animation anim = create_animation(script, TEST_ANIMATION_NAME, false);
+        REQUIRE(anim != nullptr);
+
+        int initial_cell = animation_current_cell(anim);
```
Comment (rory-cd) [inline coresdk/src/test/unit_tests/unit_test_animation.cpp:231]: This line should be removed - the value is never used.

File: coresdk/src/test/unit_tests/unit_test_animation.cpp [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+
+        // Update animation multiple times
+        for (int i = 0; i < 100; i++)
+        {
+            update_animation(anim);
+        }
+
+        // Cell may have changed after updates
+        int final_cell = animation_current_cell(anim);
+        REQUIRE(final_cell >= 0);
```
Comment (rory-cd) [inline coresdk/src/test/unit_tests/unit_test_animation.cpp:241]: This assertion would be more robust with an upper limit. For example, you could get the total frame count with:
```cpp
int total_frames = animation_frame_count(anim);
```
and use that as an upper limit.

File: coresdk/src/test/unit_tests/unit_test_animation.cpp [hunk 9]
```diff
@@ -0,0 +1,374 @@
...
+            update_animation(anim);
+        }
+
+        float time_before = animation_frame_time(anim);
+
+        // Restart animation
+        restart_animation(anim);
+
+        float time_after = animation_frame_time(anim);
+        REQUIRE(time_after == 0.0f);
```
Comment (rory-cd) [inline coresdk/src/test/unit_tests/unit_test_animation.cpp:262]: This could be changed to `REQUIRE(time_after == Approx(0.0f));` to account for any floating point inaccuracies.

File: coresdk/src/test/unit_tests/unit_test_animation.cpp [hunk 10]
```diff
@@ -0,0 +1,374 @@
...
+        REQUIRE_FALSE(animation_ended(anim));
```
Comment (rory-cd) [inline coresdk/src/test/unit_tests/unit_test_animation.cpp:263]: You could add another assertion to strengthen the test here, for example `REQUIRE(animation_current_cell(anim) == 0);`

File: coresdk/src/test/unit_tests/unit_test_animation.cpp [hunk 11]
```diff
@@ -0,0 +1,374 @@
...
+        animation anim = create_animation(script, TEST_ANIMATION_NAME, false);
+        REQUIRE(anim != nullptr);
+
+        // Update to progress the animation
+        for (int i = 0; i < 50; i++)
+        {
+            update_animation(anim);
+        }
+
+        float time_before = animation_frame_time(anim);
```
Comment (rory-cd) [inline coresdk/src/test/unit_tests/unit_test_animation.cpp:256]: `time_before` is never used. It should be utilised or removed.

File: coresdk/src/test/unit_tests/unit_test_animation.cpp [hunk 12]
```diff
@@ -0,0 +1,374 @@
...
+    SECTION("animation with percentage update")
+    {
+        animation anim = create_animation(script, TEST_ANIMATION_NAME, false);
+        REQUIRE(anim != nullptr);
+
+        // Update with percentage
+        update_animation(anim, 0.5f);
+
+        float frame_time = animation_frame_time(anim);
+        REQUIRE(frame_time > 0.0f);
```
Comment (rory-cd) [inline coresdk/src/test/unit_tests/unit_test_animation.cpp:276]: If the frame has a `frame_time` of 0, this assertion will fail. Consider `>=` rather than `>`.

This would also benefit from an upper bound, e.g. `REQUIRE(frame_time <= 12.0f);`

File: coresdk/src/test/unit_tests/unit_test_animation.cpp [hunk 13]
```diff
@@ -0,0 +1,374 @@
...
+
+        free_animation(anim);
+    }
+    SECTION("can assign animation by index")
+    {
+        animation anim = create_animation(script, TEST_ANIMATION_NAME, false);
+        REQUIRE(anim != nullptr);
+
+        int dance_idx = animation_index(script, "Dance");
+        assign_animation(anim, dance_idx);
```
Comment (rory-cd) [inline coresdk/src/test/unit_tests/unit_test_animation.cpp:307]: There is no assertion here to test the function. This could be `REQUIRE(animation_name(anim) == "dance");` or similar.

File: coresdk/src/test/unit_tests/unit_test_animation.cpp [hunk 14]
```diff
@@ -0,0 +1,374 @@
...
+        REQUIRE(animation_ended(nullptr));
+    }
+    SECTION("animation_current_cell returns 0 for null")
+    {
+        REQUIRE(animation_current_cell(nullptr) == 0);
+    }
+    SECTION("animation_name returns empty for null")
+    {
+        disable_logging(WARNING);
+        REQUIRE(animation_name(nullptr) == "");
```
Comment (rory-cd) [inline coresdk/src/test/unit_tests/unit_test_animation.cpp:353]: This could be `REQUIRE(animation_name(nullptr).empty());` for consistency with earlier usage of empty strings (line 190).

**Review by rory-cd — APPROVED** [review]
## General Information

Type of Change:
  - [x] New unit tests

This PR is a valuable addition, with unit tests covering a wide range of animation functions. Some assertions could be strengthened slightly, as noted with inline comments. 

Optional: For consistency with other tests, change all occurrences of `std::string` to `string`, and add `#include "types.h"`

## Code Quality

- [x] Repository: This Pull Request is made to the correct repository
- [x] Readability: The code is easy to read and follow
- [x] Maintainability: The code could be easily maintained or extended in the future

## Functionality

- [x] Correctness: The code meets the requirements of the task
- [x] Impact on Existing Functionality: The impact on existing functionality been considered

## Testing

- [x] Test Coverage: Unit tests provided for new or modified code
- [x] Test Results: All tests passed

## Documentation

- [x] Documentation: Applicable internal documentation is updated and clear

## Pull Request Details

- [x] PR Description: The problem being solved is clearly described
- [x] Checklist Completion: All relevant checklist items have been reviewed and completed

**Review by Broccoli811 — APPROVED** [review]
This PR is a strong and comprehensive addition to the SplashKit test suite. It meaningfully addresses the task card by converting previously manual animation checks into automated unit tests suitable for CI/CD. Coverage is thorough across animation script loading, animation creation, lifecycle behavior, and invalid/null handling.

I have run tests on my end with no apparent issues discovered. The structure and organization of the tests are clear, and the sections are particularly valuable. Using the existing kermit.txt resource keeps the tests realistic and consistent with current examples. I also appreciate the explicit testing of invalid pointers and failure paths, which improves robustness of the code.

As Rory has stated, there are a couple of improvements which can be made so ensure you follow them. Here are my minor suggestions for improvement:

- The restart test currently checks frame time and animation_ended. You could also assert that the current cell resets (returns to 0 or initial state) to fully validate restart behavior.

- Each test loads and frees kermit.txt independently. This is fine for isolation, but a small helper function (e.g. load_test_script()) at the top could reduce repetition and improve readability if more animation tests are added later.

Overall, this is a quality contribution that provides automated coverage for animations and aligns well with project testing conventions. 

Approved.
